src/databricks/labs/dqx/llm/llm_core.py

```diff
@@ -0,0 +1,315 @@
+        )
+    )
+    reasoning: str = dspy.OutputField(desc="Explanation of why these rules were chosen")
+
+
+class DQRuleGeneration(dspy.Module):
+    """
+    Generate data quality rules with improved JSON output reliability.
+
+    This class provides functionality to generate data quality rules based on schema information,
+    business descriptions, and available functions. It can optionally infer the schema from the
+    business description if schema_info is not provided or is empty.
+    """
+
+    def __init__(self, enable_schema_inference: bool = True):
+        """
+        Initialize the DQ rule generation module.
+
+        Args:
+            enable_schema_inference (bool): If True, will use SchemaGuesser when schema_info is empty.
+        """
+        super().__init__()
+        # Use Predict for reliable output
+        self.generator = dspy.Predict(RuleSignature)
+        self.enable_schema_inference = enable_schema_inference
+        self.schema_guesser = SchemaGuesser() if enable_schema_inference else None
+
+    def forward(
+        self, schema_info: str, business_description: str, available_functions: str
+    ) -> dspy.primitives.prediction.Prediction:
+        """
+        Generate data quality rules based on schema information, business descriptions, and available functions.
+
+        If schema_info is empty and enable_schema_inference is True, it will first use SchemaGuesser
+        to infer the schema from the business description.
+
+        Args:
+            schema_info (str): JSON string containing table schema with column names, types, and sample data.
+                              If empty and enable_schema_inference=True, schema will be inferred.
+            business_description (str): Natural language description of data quality requirements.
+            available_functions (str): JSON string of available DQX check functions.
+
+        Returns:
+            dspy.primitives.prediction.Prediction: A Prediction object containing the generated data quality rules,
+            reasoning, and optionally guessed_schema_json and assumptions_bullets if schema was inferred.
+        """
+        # Step 1: Infer schema if needed
+        guessed_schema_json = None
+        assumptions_bullets = None
+
+        if not schema_info or not schema_info.strip():
+            if self.enable_schema_inference and self.schema_guesser:
+                logger.info("Schema not provided, inferring from business description...")
+                schema_result = self.schema_guesser(business_description=business_description)
+                schema_info = schema_result.guessed_schema_json
+                guessed_schema_json = schema_result.guessed_schema_json
+                assumptions_bullets = schema_result.assumptions_bullets
+                logger.info(f"Inferred schema: {schema_info}")
+            else:
+                raise ValueError("schema_info is required when enable_schema_inference=False")
```

Comment:
```suggestion
                raise MissingParameterError("schema_info is required when enable_schema_inference=False")
```
from databricks.labs.dqx.errors import MissingParameterError